security/pkg/nodeagent/caclient/providers/vault/client.go

```diff
@@ -19,70 +19,135 @@ import (
+func NewVaultClientWithConfig(config string) (*VaultClient, error) {
+	params := strings.Split(config, ";")
+	if len(params) != 7 {
+		return nil, fmt.Errorf(
+			"error processing config for Vault. Expected 'backend_ca_addr;tls_cert_path;jwt_path;role;"+
+				"auth_path;sign_csr_path;ca_cert_path', but got '%s' which has %d segments", config, len(params))
+	}
+	vaultAddr := params[0]
+	tlsRootCertPath := params[1]
+	jwtPath := params[2]
+	vaultLoginRole := params[3]
+	vaultLoginPath := params[4]
+	vaultSignCsrPath := params[5]
+	vaultCACertPath := params[6]
 
-	client *api.Client
+	return NewVaultClient(
+		vaultAddr, tlsRootCertPath, jwtPath, vaultLoginRole, vaultLoginPath, vaultSignCsrPath, vaultCACertPath)
 }
 
-// NewVaultClient create a CA client for the Vault provider 1.
-func NewVaultClient(tls bool, tlsRootCert []byte,
-	vaultAddr, vaultLoginRole, vaultLoginPath, vaultSignCsrPath string) (caClientInterface.Client, error) {
-	c := &vaultClient{
-		enableTLS:        tls,
-		tlsRootCert:      tlsRootCert,
-		vaultAddr:        vaultAddr,
-		vaultLoginRole:   vaultLoginRole,
-		vaultLoginPath:   vaultLoginPath,
-		vaultSignCsrPath: vaultSignCsrPath,
+// NewVaultClient creates a CA client for the Vault PKI.
+func NewVaultClient(vaultAddr, tlsRootCertPath, jwtPath, loginRole, loginPath,
+	signCsrPath, caCertPath string) (*VaultClient, error) {
+	c := &VaultClient{
+		enableTLS:       true,
+		vaultAddr:       vaultAddr,
+		tlsRootCertPath: tlsRootCertPath,
+		jwtPath:         jwtPath,
+		loginRole:       loginRole,
+		loginPath:       loginPath,
+		signCsrPath:     signCsrPath,
+		caCertPath:      caCertPath,
+	}
+	if strings.HasPrefix(c.vaultAddr, "http:") {
+		c.enableTLS = false
 	}
 
-	var client *api.Client
+	jwtLoader, tlErr := util.NewJwtLoader(c.jwtPath)
+	if tlErr != nil {
+		return nil, fmt.Errorf("failed to create token loader to load the tokens: %v", tlErr)
+	}
+	stopCh := make(chan struct{})
+	go jwtLoader.Run(stopCh)
+	c.jwtLoader = jwtLoader
+
+	var client *vaultapi.Client
 	var err error
-	if tls {
-		client, err = createVaultTLSClient(vaultAddr, tlsRootCert)
+	if c.enableTLS {
+		client, err = createVaultTLSClient(c.vaultAddr, c.tlsRootCertPath)
 	} else {
-		client, err = createVaultClient(vaultAddr)
+		client, err = createVaultClient(c.vaultAddr)
 	}
 	if err != nil {
 		return nil, err
 	}
 	c.client = client
-	vaultClientLog.Infof("created Vault client for Vault address: %s, TLS: %v", vaultAddr, tls)
-
-	return c, nil
-}
 
-// CSR Sign calls Vault to sign a CSR.
-func (c *vaultClient) CSRSign(ctx context.Context, reqID string, csrPEM []byte, saToken string,
-	certValidTTLInSec int64) ([]string /*PEM-encoded certificate chain*/, error) {
-	token, err := loginVaultK8sAuthMethod(c.client, c.vaultLoginPath, c.vaultLoginRole, saToken)
+	token, err := loginVaultK8sAuthMethod(c.client, c.loginPath, c.loginRole, jwtLoader.GetJwt())
```

Comment:
As loginVaultK8sAuthMethod() will be called in CSRSign(), is it needed to call loginVaultK8sAuthMethod() in NewVaultClient()?

Reply:
Yes I think it's good to have it. This benefits in that it raises early alert if the authentication doesn't work.